crates/bevy_pbr/src/render/fog.wgsl

```diff
@@ -0,0 +1,67 @@
+#define_import_path bevy_pbr::fog
+
+// Fog formulas adapted from:
+// https://learn.microsoft.com/en-us/windows/win32/direct3d9/fog-formulas
+// https://catlikecoding.com/unity/tutorials/rendering/part-14/
+
+fn linear_fog(
+    input_color: vec4<f32>,
+    distance: f32,
+    scattering: f32,
+) -> vec4<f32> {
+    var result = fog.base_color;
+    if (scattering > 0.0) {
+        result = vec4<f32>(mix(result.rgb, fog.scattering_color.rgb, min(scattering * fog.scattering_color.a, 1.0)), result.a);
+    }
+    let start = fog.be.x;
+    let end = fog.be.y;
+    result.a *= 1.0 - clamp((end - distance) / (end - start), 0.0, 1.0);
+    return vec4<f32>(mix(input_color.rgb, result.rgb, result.a), input_color.a);
+}
+
+fn exponential_fog(
+    input_color: vec4<f32>,
+    distance: f32,
+    scattering: f32,
+) -> vec4<f32> {
+    var result = fog.base_color;
+    if (scattering > 0.0) {
+        result = vec4<f32>(mix(result.rgb, fog.scattering_color.rgb, min(scattering * fog.scattering_color.a, 1.0)), result.a);
+    }
+    let density = fog.be.x;
+    result.a *= 1.0 - 1.0 / exp(distance * density);
+    return vec4<f32>(mix(input_color.rgb, result.rgb, result.a), input_color.a);
+}
+
+fn exponential_squared_fog(
+    input_color: vec4<f32>,
+    distance: f32,
+    scattering: f32,
+) -> vec4<f32> {
+    var result = fog.base_color;
+    if (scattering > 0.0) {
+        result = vec4<f32>(mix(result.rgb, fog.scattering_color.rgb, min(scattering * fog.scattering_color.a, 1.0)), result.a);
+    }
+    let density = fog.be.x;
+    result.a *= 1.0 - 1.0 / exp(pow(distance * density, 2.0));
+    return vec4<f32>(mix(input_color.rgb, result.rgb, result.a), input_color.a);
+}
+
+// Fog formula adapted from:
+// https://iquilezles.org/articles/fog/
+
+fn atmospheric_fog(
+    input_color: vec4<f32>,
+    distance: f32,
+    scattering: f32,
+) -> vec4<f32> {
+    var result = fog.base_color;
+    if (scattering > 0.0) {
+        result = vec4<f32>(mix(result.rgb, fog.scattering_color.rgb, min(scattering * fog.scattering_color.a, 1.0)), result.a);
+    }
+
+    let extinction_color = 1.0 - 1.0 / vec3<f32>(exp(distance * fog.be.r), exp(distance * fog.be.g), exp(distance * fog.be.b));
+    let inscattering_color = 1.0 - 1.0 / vec3<f32>(exp(distance * fog.bi.r), exp(distance * fog.bi.g), exp(distance * fog.bi.b));
```

Comment:
Unfortunately I don't think this will work, since $k \cdot e^{x} \ne e^{k \cdot x}$.

Reply:
The wgsl `*` of two vectors is component-wise multiplication and exp() on a vector is also component-wise. So changing vec3(exp(v.x), exp(v.y), exp(v.z))

Reply:
Oh, I think you accidentally had the parenthesis in the wrong spot. In that case you mean:

```suggestion
    let extinction_color = 1.0 - 1.0 / exp(vec3<f32>(distance * fog.be));
    let inscattering_color = 1.0 - 1.0 / exp(vec3<f32>(distance * fog.bi));
```

That should indeed work. I didn't think of using per-component `exp()`. Let me tweak the code to use it.

Reply:
No. The parentheses are in the right position as far as I can see. `distance` is a scalar. `vec3<f32>(distance)` makes it into a 3D vector with the same value for each component. This vector value can be reused for both `be` and `bi` computations. And `*` multiplies vector component-wise in wgsl and other shader languages, it isn't a dot/scalar product.

Reply:
Ouch, I misread GitHub's diff because of the highlighting it applies, I read it as if the pairing was between `exp(` and the highlighted parameter `)` 🤦. You're 100% right.

<img width="284" alt="image" src="https://user-images.githubusercontent.com/418473/199882993-e509696b-0801-4c44-968b-166b8b4d2952.png">

Reply:
I did some experiments and it looks like it's okay with splatting the entire thing implicitly, so this works, which is even shorter without the explicit `vec3<f32>`:

```
    let extinction_color = 1.0 - 1.0 / exp(distance * fog.be);
    let inscattering_color = 1.0 - 1.0 / exp(distance * fog.bi);
```